Blog post: New conversion from cgroup v1 CPU shares to v2 CPU weight (#52793)
kubernetes/website: content/en/blog/_posts/2025-10-25-new-cgroup-v1-to-v2-conversion-formula/index.md

---
layout: blog
title: 'New Conversion from cgroup v1 CPU Shares to v2 CPU Weight'
date: 2025-10-25T05:00:00-08:00
draft: true
math: true
slug: new-cgroup-v1-to-v2-cpu-conversion-formula
author: >
   [Itamar Holder](https://github.com/iholder101) (Red Hat)
---

I'm excited to announce the implementation of an improved conversion formula 
from cgroup v1 CPU shares to cgroup v2 CPU weight. This enhancement addresses 
critical issues with CPU priority allocation for Kubernetes workloads when 
running on systems with cgroup v2.

## Background

Kubernetes was originally designed with cgroup v1 in mind, where CPU shares 
were defined simply by assigning the container's CPU requests in millicpu 
form. 

> For example, a container requesting 1 CPU (1024m) would get \(cpu.shares = 1024\).

Reviewer: created issue https://github.com/kubernetes/website/issues/54411

> After a while, cgroup v1 was stared being replaced by its successor, 

Reviewer 2: ```suggestion
After a while, cgroup v1 started being replaced by its successor, 
```

cgroup v2. In cgroup v2, the concept of CPU shares (which ranges from 2 to 
262144, or from 2¹ to 2¹⁸) was replaced with CPU weight (which ranges from 
[1, 10000], or 10⁰ to 10⁴.

With the transition to cgroup v2, 
[KEP-2254](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2) 
introduced a conversion formula to map cgroup v1 CPU shares to cgroup v2 CPU 
weight. The conversion formula was defined as: `cpu.weight = (1 + ((cpu.shares - 2) * 9999) / 262142)`

This formula linearly maps between [2¹, 2¹⁸] to [10⁰, 10⁴].

![Linear conversion formula](./linear-conversion.png "formula")

While this approach is simple, the linear mapping imposes a few significant 
problems and impacts both performance and configuration granularity.

## Problems with Previous Conversion Formula

The current conversion formula creates two major issues:

> ### 1. Reduced Priority Against Non-Kubernetes Workloads

Reviewer 2: ```suggestion
### 1. Reduced priority against non-Kubernetes workloads
```

In cgroup v1, the default CPU shares is `1024`, meaning a container 
requesting 1 CPU has equal priority with system processes that live outside 
of Kubernetes' scope.
However, in cgroup v2, the default CPU weight is `100`, but the current 
formula converts 1 CPU (1024m) to only `≈39` weight - less than 40% of the 
default.

**Example:**
- Container requesting 1 CPU (1024m)
- cgroup v1: `cpu.shares = 1024` (equal to default)
- cgroup v2 (current): `cpu.weight = 39` (much lower than default 100)

This means that after moving to cgroup v2, Kubernetes (or, OCI) workloads would 

> de-factor reduce their CPU priority against non-Kubernetes processes. The 

Reviewer 2: I'm not sure if this is intended to be "de facto reduce" or if it's just a term I'm unfamiliar with as written, if you wouldn't mind double checking?

With plans to allow [writable cgroups for unprivileged containers](https://github.com/kubernetes/enhancements/issues/5474),
this becomes even 
more relevant.

## New Conversion Formula

### Description
The new formula is more complicated, but does a much better job mapping 
between cgroup v1 CPU shares and cgroup v2 CPU weight:

```math
cpu.weight = \lceil 10^{(L^{2}/612 + 125L/612 - 7/34)} \rceil, \text{ where: } L = \log_2(cpu.shares)
```

The idea is that this is a quadratic function to cross the following values:
- (2, 1): The minimum values for both ranges.
- (1024, 100): The default values for both ranges.
- (262144, 10000): The maximum values for both ranges.

Visually, the new function looks as follows:

![2025-10-25-new-cgroup-v1-to-v2-conversion-formula-new-conversion.png](./new-conversion-formula.png)

And if you zoom in to the important part:

![2025-10-25-new-cgroup-v1-to-v2-conversion-formula-new-conversion-zoom.png](./new-conversion-formula-zoom.png)

The new formula is "close to linear", yet it is sophistically designed to 
map the ranges in a clever way so the three important points above would 
cross.

> ### How It Solves the Problems

Reviewer 2: ```suggestion
### How it solves the problems
```

> 1. **Better Priority Alignment:**

Reviewer 2: ```suggestion
1. **Better priority alignment:**
```
